clang/lib/Sema/SemaBounds.cpp

```diff
@@ -5623,6 +5712,84 @@ namespace {
       return Pair;
     }
 
+    // SynthesizeMembers modifies the set AbstractSets to include AbstractSets
+    // for member expressions whose target bounds use the value of the member
+    // expression M that is being modified via an assignment.
+    void SynthesizeMembers(Expr *E, MemberExpr *M, CheckedScopeSpecifier CSS,
+                           AbstractSetSetTy &AbstractSets) {
+      if (!E)
+        return;
+      Lexicographic Lex(S.Context, nullptr);
+      E = Lex.IgnoreValuePreservingOperations(S.Context, E->IgnoreParens());
+
+      // Recursively synthesize AbstractSets for certain subexpressions of E.
+      switch (E->getStmtClass()) {
+        case Expr::MemberExprClass:
+          break;
+        case Expr::ImplicitCastExprClass:
+        case Expr::CStyleCastExprClass: {
+          CastExpr *CE = cast<CastExpr>(E);
+          SynthesizeMembers(CE->getSubExpr(), M, CSS, AbstractSets);
+          return;
+        }
+        case Expr::UnaryOperatorClass: {
+          UnaryOperator *UO = cast<UnaryOperator>(E);
+          SynthesizeMembers(UO->getSubExpr(), M, CSS, AbstractSets);
+          return;
+        }
+        case Expr::ArraySubscriptExprClass: {
+          ArraySubscriptExpr *AE = cast<ArraySubscriptExpr>(E);
+          SynthesizeMembers(AE->getBase(), M, CSS, AbstractSets);
+          return;
+        }
+        case Expr::BinaryOperatorClass: {
+          BinaryOperator *BO = cast<BinaryOperator>(E);
+          Expr *LHS = BO->getLHS();
+          Expr *RHS = BO->getRHS();
+          if (LHS->getType()->isPointerType())
+            SynthesizeMembers(LHS, M, CSS, AbstractSets);
+          if (RHS->getType()->isPointerType())
+            SynthesizeMembers(RHS, M, CSS, AbstractSets);
+          return;
+        }
+        default:
+          return;
+      }
+
+      MemberExpr *ME = dyn_cast<MemberExpr>(E);
+      if (!ME)
+        return;
+
+      // Synthesize AbstractSets for a member expression `Base->Field`
+      // or `Base.Field`.
+      Expr *Base = ME->getBase();
+      FieldDecl *Field = dyn_cast<FieldDecl>(ME->getMemberDecl());
+      if (!Field)
+        return;
+
+      RecordDecl *Parent = Field->getParent();
+      if (!Parent)
+        return;
+
+      // Iterate over each sibling field F of Field to check whether the
+      // target bounds of `Base->F` or `Base.F` use the value of M.
+      for (auto It = Parent->field_begin(), E = Parent->field_end(); It != E; ++It) {
```

Comment:
Once PR #1047 is merged, we can iterate over the fields in `BoundsSiblingFields[Field]` instead of all the fields in `Parent`. `BoundsSiblingFields[Field]` is the set of fields in `Parent` in whose declared bounds `Field` appears. If a field `F` in `Parent` is not in `BoundsSiblingFields[Field]`, then the target bounds of `Base->F` (or `Base.F`) do not use the value of `M`, so we do not need to create a `MemberExpr *` for `Base` and `F`, concretize or expand its declared bounds, or create an `AbstractSet *` for `Base` and `F`.